rclpy/src/rclpy/_rclpy_qos_event.c

```diff
@@ -0,0 +1,334 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <rcl/event.h>
+
+typedef union _qos_event_callback_data {
+  // Subscription events
+  rmw_requested_deadline_missed_status_t requested_deadline_missed;
+  rmw_liveliness_changed_status_t liveliness_changed;
+  // Publisher events
+  rmw_offered_deadline_missed_status_t offered_deadline_missed;
+  rmw_liveliness_lost_status_t liveliness_lost;
+} _qos_event_callback_data_t;
+
+typedef PyObject * (* _qos_event_data_filler_function)(_qos_event_callback_data_t *);
+
+bool
+_check_rcl_return(rcl_ret_t ret, const char * error_msg)
+{
+  if (ret != RCL_RET_OK) {
+    PyErr_Format(PyExc_RuntimeError,
+      "%s: %s", error_msg, rcl_get_error_string().str);
+    rcl_reset_error();
+    return false;
+  }
+  return true;
+}
+
+void
+_destroy_event_capsule(PyObject * pycapsule)
+{
+  rcl_event_t * event = (rcl_event_t *)PyCapsule_GetPointer(pycapsule, "rcl_event_t");
+  rcl_ret_t ret = rcl_event_fini(event);
+  PyMem_Free(event);
+  _check_rcl_return(ret, "Failed to fini 'rcl_event_t'");
```

Comment:
Please avoid raising exceptions in pycapsule destructors. If the capsule is garbage collected this results in tracebacks in random places. Instead I recommend following the pattern used by the subscription where it warns if an error occurs.

https://github.com/ros2/rclpy/blob/897123438ca05834e6ed943045d871c568c26448/rclpy/src/rclpy/_rclpy.c#L1671-L1695

Reply:
Ok I will use this warning pattern.